_posts/2019-12-17-zero-to-main-rust-1.md

```diff
@@ -0,0 +1,433 @@
+```
+
+This is called a "global attribute", and it is stating that this Rust code will not be using the Rust Standard Library. Attributes in Rust are sometimes used similarly to `#pragma` in C, in that they can change certain behaviors of the compiler.
+
+Rust provides a number of built-in library components, but the two main ones are:
+
+1. The Rust Standard Library
+2. The Rust Core Library
+
+While the Standard Library contains a number of useful components, such as data structures, and interfaces for opening files and sockets, it generally requires your target to have these things! For bare metal applications, we can instead forego this library, and only use the Rust Core Library, does not have these requirements.
+
+Rust as a language has a concept of "modules", which can be used to organize and provide namespaces for your code. Libraries or applications in Rust are called "Crates", and each has it's own namespace. This is why we saw the symbol `from_scratch::reset_handler` in our linker script: It was referring to the `reset_handler` function in the `from_scratch` crate (which is the crate in this example).
+
+To use items from another crate, including the `core` library, you can import these items in using the `use` syntax:
+
+```rust
+use core::{
+    mem::zeroed,
+    panic::PanicInfo,
+    ptr::{read, write_volatile},
+};
+```
+
+This imports the symbols into the current context so that they can be used. Most symbols in Rust are not available in a global namespace, which helps to avoid naming collisions.
+
+However in some cases, it is important to have a globally defined symbol. As part of the ABI of the Cortex-M platform, we need to provide the address of the reset handler in a very specific location. Let's look at how we do that in Rust:
+
+### Setting the Reset Vector
+
+```rust
+#[link_section = ".vector_table.reset_vector"]
+#[no_mangle]
+pub static __RESET_VECTOR: fn() -> ! = reset_handler;
+```
+
+Let's unpack that from the bottom up:
+
+```rust
+pub static __RESET_VECTOR: fn() -> ! = reset_handler;
+```
+
+This line defines a symbol called `__RESET_VECTOR` at static scope. The type of this symbol is `fn() -> !`, which is a pointer to a function that never returns, or that "diverges". The value of this symbol is `reset_handler`, which is the name of a function in our program. Functions are a first class items in Rust (similar to Python), so we can use the names of functions as a value that represents a function pointer.
+
+```rust
+#[no_mangle]
+```
+
+This is another attribute, like our `#![no_std]`. By starting with `#[` instead of `#![`, we can tell this is a local attribute instead of a global attribute, which means it only applies to the next item, instead of the whole module.
+
+The `#[no_mangle]` attribute tells the compiler **not** to mangle this symbol, so it will show up as `__RESET_VECTOR`, not `from_scratch::__RESET_VECTOR`. This is important, so that our linker script can understand the symbol we are providing.
```

Comment:
Add a sentence about what name mangling is, most C developers aren't familiar (C++ concept)

Reply:
Will do! I intended for my `__RESET_VECTOR` vs `from_scratch::__RESET_VECTOR` line to cover this, without going into details around special characters in linker symbols, or concepts like generics, but I can give a bit more context.